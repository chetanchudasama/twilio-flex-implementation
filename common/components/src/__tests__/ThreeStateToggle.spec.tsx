/* eslint-disable react/jsx-props-no-spreading */
import { createMuiTheme, MuiThemeProvider } from "@material-ui/core";
import { mount } from "enzyme";
import flushPromise from "flush-promises";
import React from "react";
import { create } from "react-test-renderer";
import {
  ThreeStateToggle,
  ThreeStateToggleProps,
} from "../ThreeStateToggle/ThreeStateToggle";

const core: ThreeStateToggleProps = {
  checked: null,
  onChange: () => jest.fn(),
};
let props: ThreeStateToggleProps = { ...core };

const theme = createMuiTheme({
  typography: {
    useNextVariants: true,
  },
});

describe("ThreeStateToggle", () => {
  beforeEach(() => {
    props = { ...core };
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  it("when toggle is initialized with checked state, renders correctly", () => {
    props.checked = true;
    const wrapper = create(
      <MuiThemeProvider theme={theme}>
        <ThreeStateToggle {...props} />
      </MuiThemeProvider>
    ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("when toggle is initialized with unchecked state, renders correctly", () => {
    props.checked = false;
    const wrapper = create(
      <MuiThemeProvider theme={theme}>
        <ThreeStateToggle {...props} />
      </MuiThemeProvider>
    ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
  it("when toggle is initialized with intermediate state, renders correctly", () => {
    props.checked = null;
    const wrapper = create(
      <MuiThemeProvider theme={theme}>
        <ThreeStateToggle {...props} />
      </MuiThemeProvider>
    ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("when clicked on left button, change event should be triggered with false flag", async () => {
    const onChange = jest.fn();
    props.onChange = onChange;
    props.checked = null;

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <ThreeStateToggle {...props} />
      </MuiThemeProvider>
    );
    await flushPromise();

    wrapper.find(".btn-no").last().simulate("click");
    await flushPromise();

    expect(onChange).toHaveBeenCalledWith(false);
  });

  it("when clicked on right button, change event should be triggered with true flag", async () => {
    const onChange = jest.fn();
    props.onChange = onChange;
    props.checked = null;

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <ThreeStateToggle {...props} />
      </MuiThemeProvider>
    );
    await flushPromise();

    wrapper.find(".btn-yes").last().simulate("click");
    await flushPromise();

    expect(onChange).toHaveBeenCalledWith(true);
  });

  it("when clicked on default button, change event should be triggered with null flag", async () => {
    const onChange = jest.fn();
    props.onChange = onChange;
    props.checked = true;

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <ThreeStateToggle {...props} />
      </MuiThemeProvider>
    );
    await flushPromise();

    wrapper.find(".btn-intermediate").last().simulate("click");
    await flushPromise();

    expect(onChange).toHaveBeenCalledWith(null);
  });
});
