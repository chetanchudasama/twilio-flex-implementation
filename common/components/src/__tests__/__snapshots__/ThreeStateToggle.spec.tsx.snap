// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeStateToggle when toggle is initialized with checked state, renders correctly 1`] = `
<div
  className="MuiToggleButtonGroup-root-5 ThreeStateToggleContent-root-1 MuiToggleButtonGroup-selected-6"
>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 btn-intermediate"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={2}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      Not Asked
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 btn-no"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={1}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      No
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 MuiToggleButton-selected-9 ThreeStateToggleContent-selected-3 btn-yes"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={0}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      Yes
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
</div>
`;

exports[`ThreeStateToggle when toggle is initialized with intermediate state, renders correctly 1`] = `
<div
  className="MuiToggleButtonGroup-root-5 ThreeStateToggleContent-root-1 MuiToggleButtonGroup-selected-6"
>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 MuiToggleButton-selected-9 ThreeStateToggleContent-selected-3 btn-intermediate"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={2}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      Not Asked
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 btn-no"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={1}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      No
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 btn-yes"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={0}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      Yes
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
</div>
`;

exports[`ThreeStateToggle when toggle is initialized with unchecked state, renders correctly 1`] = `
<div
  className="MuiToggleButtonGroup-root-5 ThreeStateToggleContent-root-1 MuiToggleButtonGroup-selected-6"
>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 btn-intermediate"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={2}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      Not Asked
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 MuiToggleButton-selected-9 ThreeStateToggleContent-selected-3 btn-no"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={1}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      No
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
  <button
    className="MuiButtonBase-root-11 MuiToggleButton-root-7 ThreeStateToggleContent-button-2 btn-yes"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
    value={0}
  >
    <span
      className="MuiToggleButton-label-10 ThreeStateToggleContent-label-4"
    >
      Yes
    </span>
    <span
      className="MuiTouchRipple-root-14"
    />
  </button>
</div>
`;
